kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/throwables/CovariantThrowableHandling.kt

```diff
@@ -260,3 +260,32 @@ inline fun <reified T : Throwable> shouldNotThrow(block: () -> Any?) {
       throw failure("No exception expected, but a ${thrown::class.simpleName} was thrown.", thrown)
    throw thrown
 }
+
+/**
+ * Verifies if a block of code will either throw an [AssertionError] or no exception at all. It can only be invoked inside [assertSoftly] blocks.
+ *
+ * Use this function to wrap a block of code that might throw an [AssertionError]. Use it to wrap invocations of non-kotest functions
+ * such as [verify] from mockk library, so that they can be invoked inside [assertSoftly] blocks, and
+ * errors are only thrown at the end of the block, rather than immediately. Therefore, the signature cannot return
+ * the throwable type, as in the case of a failure, there would be neither an immediate throws, nor a type to return.
+ */
+inline fun shouldPass(block: () -> Any?) {
```

Comment:
This function is very similar to `shouldNotThrow<> { ... }`

Perhaps we should either reuse the method or have this one have a similar name as  `shouldNotThrow`. This way we don't have 2 similar block code lying around

Reply:
> This function is very similar to `shouldNotThrow<> { ... }`
> 
> Perhaps we should either reuse the method or have this one have a similar name as `shouldNotThrow`. This way we don't have 2 similar block code lying around

I concur it is similar, but unfortunately it won't do what I want:

```kotlin
class VerifyTest: StringSpec() {
    private val mockService = mockk<MyService>(relaxed = true)

    init {
        "try with shouldNotThrow" {
            assertSoftly {
                shouldNotThrow<AssertionError> {
                    verify(exactly = 1) { mockService.doSomething(any()) }
                }
                (2 + 2) shouldBe 5
            }
        }
    }
}

class MyService {
    fun doSomething(params: String) {}
}

--- the output

No exception expected, but a AssertionError was thrown.
java.lang.AssertionError: No exception expected, but a AssertionError was thrown.

(snip)

Caused by: java.lang.AssertionError: Verification failed: call 1 of 1: MyService(#1).doSomething(any())) was not called

```

There are two issues here:
* `(2 + 2) shouldBe 5` is not run at all
* Unnecessary noise here: "No exception expected, but a AssertionError was thrown." - ideally all we need is "Verification failed: call 1 of 1: MyService(#1).doSomething(any())) was not called"

Can you suggest a better name instead of `shouldPass`?

I concur the following code is very similar:

```kotlin
   assertionCounter.inc()
   val thrownThrowable = try {
      block()
      null
   } catch (thrown: Throwable) {
      thrown
   }
```
but it is not exactly the same. Let me try reusing code.

Reply:
IMO it's more similar to `shouldThrowSoftly`

Reply:
> This function is very similar to `shouldNotThrow<> { ... }`
> 
> Perhaps we should either reuse the method or have this one have a similar name as `shouldNotThrow`. This way we don't have 2 similar block code lying around

I've extracted some common code into a function which is reused

Reply:
What Leo said was the first thing that sprung to mind for me as well. Perhaps we should take some time to discuss further before merging? 

As a user of Kotest, I feel that `shouldPass` and `shouldNotThrowAny { ... }`  are semantically indifferent and it's confusing why both would exist.

A possible workaround could be to somehow create a DSL-scope within `assertSoftly` and only expose the `shouldPass` API there?

Reply:
> What Leo said was the first thing that sprung to mind for me as well. Perhaps we should take some time to discuss further before merging?
> 
> As a user of Kotest, I feel that `shouldPass` and `shouldNotThrowAny { ... }` are semantically indifferent and it's confusing why both would exist.
> 
> A possible workaround could be to somehow create a DSL-scope within `assertSoftly` and only expose the `shouldPass` API there?

I concur that it would be natural to use `shouldNotThrowAny` in this use case. But it won't do what we want at this time, as it isn't compatible with `assertSoftly`. 

So how about the following change: make all the functions like `shouldNotThrowAny` , `shouldNotThrow`, `shouldThrowAny` etc compatible with `assertSoftly`. Also deprecate `shouldThrowSoftly` - right now it's the only function that works with `assertSoftly`. That done, we shall remove `shouldPass`, it won't be needed anymore. WDYT @Kantis